src/Microsoft.Data.Analysis/DataFrame.cs

```diff
@@ -597,15 +597,45 @@ private void OnColumnsChanged()
         private DataFrame Sort(string columnName, bool isAscending)
         {
             DataFrameColumn column = Columns[columnName];
-            DataFrameColumn sortIndices = column.GetAscendingSortIndices();
```

Comment:
Would it be better if we fixed the `GetAscendingSortIndices()` method itself? I see there is a comment in it saying "The return sortIndices contains only the non-null indices". Maybe we could add an option to that method telling it to return `null` indices or not.

Reply:
sure thing @eerhardt , 

I think my initial reason for not changing it was because the .Median method used it and i think i came across another problem (That doesn't spring to mind at the moment). But what i will do is look over this issue again and see improve the commit based on your review.

Thanks,
Ramon

Reply:
Actually, I don't think we need to change `GetAscendingSortIndices`, but rather `Sort` itself. The bug is this call `oldColumn.Clone(sortIndices, !isAscending, oldColumn.NullCount);`. `sortIndices` here is guaranteed to be either less than or equal to `oldColumn.Length` in size. So, if we replace `oldColumn.NullCount` with `oldColumn.Length - sortIndices.Length`, that should fix the bug :)

Reply:
@pgovind that works! I will update my commit and just add a new test case :).

Reply:
Hi @pgovind ,
On closer inspection the issue with the solution you mentioned is that Nulls are appended to the end of columns where it isn't needed. 

I created a testcase called TestOrderByWithNulls  in my new commit which fails because the null counts will no longer match between the old and new columns with your proposed solution. (Since the nullCount parameter will append null values inplace of actual values). 

Apologies for the lack of oversight on my side for pushing this new commit in the first place, I really shouldn't have made the assumption there that it worked. 

Do let me know if you have any other solutions? I can look at ways to improve my first commit which worked but needs improvement.

Kind Regards,
Ramon

Reply:
Ah indeed, I didn't consider that case fully in my suggestion. I spent some time looking at this today and this is my suggestion. We need to change the `GetAscendingSortIndices` method like @eerhardt suggests to return the null indices in a column in addition to the sort indices. We also need to create a similar `GetDescendingSortIndices` method. Here is what I would do:
1) Change the current `GetSortIndices` method to look like:
```
private PrimitiveDataFrameColumn<long> GetSortIndices(IComparer<T> comparer, out PrimitiveDataFrameColumn<long> columnNullIndices)
```
If we look at the implementation of `GetSortIndices` in PrimitiveDataFrameColumn.Sort.cs,  after the `IntrospectiveSort` call, we have a for loop to only populate `nonNullSortIndices`. I would write an else condition here and populate the `columnNullIndices` (after `new`-ing it) here like : `columnNullIndices.Append(sortIndices[i] + b * _columnContainer.Buffers[0].Length);`. The `GetSortIndices` method must also now return `columnSortIndices`.
2) Now that `GetSortIndices` is changed, we need to modify its callers, i.e. `GetAscendingSortIndices`. The return value of the `GetSortIndices` call are the `sortIndices` (which still only contain non-null values). The out parameter in `GetSortIndices` contains the `nullIndices`. A simple for loop that appends `nullIndices[i]` to `sortIndices` would suffice. At this point, for an ascending sort call, we have the full `sortIndices` column and don't have to deal with adding `nulls`.

Now we deal with callers of `GetAscendingSortIndices`: 
3) In `DataFrame.Sort` in `DataFrame.cs`, we should simply change the `old.Column.Clone(sortIndices, ...)` to just `oldColumn.Clone(sortIndices)`.
4) Similar change in `DataFrameColumn.Sort` in `DataFrameColumn.cs`.
5) Similar change in `PrimitiveDataFrameColumn.Sort` in `PrimitiveDataFrameColumn.cs`.
6) For `PrimitiveDataFrameColumn.Median`, we probably need to calculate `middle` now as :
```
// The last NullCount indices in sortIndices need to be ignored
            long middle = (sortIndices.Length - NullCount) / 2;
```
At this point, all the asserts associated with `var sortedDf = df.OrderBy("Int");` in the new unit test should pass. 
Then, we need to repeat all the steps above for `StringDataFrameColumn.cs` (except `median` of course) so that we can sort in ascending order via a `StringDataFrameColumn`. This should let all the asserts under `sortedDf = df.OrderBy("String");` in the unit test pass. 

Finally, we can leverage these changes to make a new `GetDescendingSortIndices` API. 
1) Add a new internal `GetDescendingSortIndices` API on `DataFrameColumn` (similar to the existing `GetAscendingSortIndices` virtual method.
2) Override the `GetDescendingSortIndices` method in `PrimitiveDataFrameColumn.Sort.cs`. Something like the following would work:
``` csharp
        internal override PrimitiveDataFrameColumn<long> GetDescendingSortIndices()
        {
            Comparison<T> descendingComparison = new Comparison<T>((T a, T b) =>
            {
                return Comparer<T>.Default.Compare(b, a);
            });
            Comparer<T> descending = Comparer<T>.Create(descendingComparison);
            // The return sortIndices contains only the non null indices. 
            PrimitiveDataFrameColumn<long> sortIndices = GetSortIndices(descending, out PrimitiveDataFrameColumn<long> nullIndices);
            for (long i = 0; i < nullIndices.Length; i++)
            {
                sortIndices.Append(nullIndices[i]);
            }
            return sortIndices;
        }
```
This has similar behavior to ascending sort in that it sorts the valid values in descending order and adds the `null` values to the end of the column. 
Then, we'd need to change
3) Both `PrimitiveDataFrameColumn.Sort` and `StringDataFrameColumn`: A simple `if (ascending)` condition will let us call `GetAscendingSortIndices` or `GetDescendingSortIndices` appropriately. Nothing else needs to be changed here I think.
4) Similar change in `DataFrameColumn.Sort` in `DataFrameColumn.cs`. 
5) Similar change in `DataFrame.Sort` in `DataFrame.cs`. 

This should enable all the descending sort tests in the new unit test pass.

I think I covered all the cases to consider, but let me know if I missed something! I would recommend getting the ascending sort changes working first for `PrimitiveDataFrameColumn` -> Ascending sort changes for `StringDataFrameColumn` -> Descending sort changes in the same order. Let me know if you need help!! I can answer questions or review changes :)

Reply:
Hi @pgovind ,
Much appreciated for the detailed response/solution. I have implemented your changes and it works fine! 

> 3) In DataFrame.Sort in DataFrame.cs, we should simply change the old.Column.Clone(sortIndices, ...) to just oldColumn.Clone(sortIndices).

Just an FYI I followed the above for all Sort methods.

Kind Regards,
Ramon